craft_parts/executor/collisions.py

```diff
@@ -28,22 +28,60 @@
 from craft_parts.permissions import Permissions, permissions_are_compatible
 
 
-def check_for_stage_collisions(part_list: List[Part]) -> None:
+def check_for_stage_collisions(
+    part_list: List[Part], partitions: Optional[List[str]]
+) -> None:
     """Verify whether parts have conflicting files to stage.
 
-    :param part_list: The list of parts to be tested.
+    If the partitions feature is enabled, then check if parts have conflicting files to
+        stage for each partition.
+    If the partitions feature is disabled, only check for conflicts in the default
+        stage directory.
+
+    :param part_list: The list of parts to check.
+    :param partitions: An optional list of partition names.
+
+    :raises PartConflictError: If conflicts are found.
+    :raises FeatureError: If partitions are specified but the feature is not enabled or
+        if partitions are not specified and the feature is enabled.
+    """
+    if partitions and not Features().enable_partitions:
+        raise errors.FeatureError(
+            "Partitions specified but partitions feature is not enabled."
+        )
+
+    if partitions is None and Features().enable_partitions:
+        raise errors.FeatureError(
+            "Partitions feature is enabled but no partitions specified."
+        )
+
+    if partitions:
```

Comment:
This can be written as

```python
 for partition in partitions or [None]:
  _check_for_stage_collisions...
```

but requires silencing pyright.